# ***Auto Correct / Spelling Checker***
<br>
A NLP based project based on nltk Corpus and Shakesphere Poems. The Modal correct all the spellings which are one edit wrong. <br>
Edit : one of the following from Delete, Insert, Replace, Swap <br>

## Timeline
<br>
 0. Importing all Libraries <br>
 1. Creating Vocabalary <br>
 2. Processing the Corpus Data <br>
 3. Creation of a Word Counter and Probibility Matrix <br>
 4. Creating the Edit Functions - Delete,Insert,Swap,Replace <br>
 5. Creating the Modal and Min Edit Algorithm <br>

## *Importing all Libraries*

In [237]:
import nltk
import numpy as np
import pandas as pd
nltk.download('words')
from nltk.corpus import words
import re

[nltk_data] Downloading package words to
[nltk_data]     C:\Users\SHALIN\AppData\Roaming\nltk_data...
[nltk_data]   Package words is already up-to-date!


## *Creating Vocabalary*

In [18]:
with open('vocab.txt') as f:
    text = f.read()
text_lower = text.lower()
vocab_1 = re.findall(r'\w+',text_lower)

In [25]:
vocab_2 = words.words()

***Formatting Raw Data***

In [28]:
with open('shakespeare.txt') as f:
    text = f.read()
all_words = re.findall(r'\w+',text_lower)

In [29]:
vocab_3 = set(all_words) 

In [30]:
vocab_3 = list(vocab_3)

In [32]:
complete_vocab = vocab_1 + vocab_2 + vocab_3

In [33]:
complete_vocab = set(complete_vocab)
complete_vocab = list(complete_vocab)

In [34]:
len(complete_vocab)

246211

#### So we have a vocablary of 246211 words

## *Words_Processing and Creating Counter*

In [36]:
word_count = {}

In [37]:
for word in complete_vocab:
    word_count[word] = 1

In [39]:
for word in all_words:
    word_count[word] += 1

In [53]:
for word in vocab_2:
    word_count[word] += 10

In [54]:
for word in vocab_1:
    word_count[word] += 5

In [104]:
n = 0
for word in complete_vocab:
    n += word_count[word]

In [105]:
n

2932315

## *Probabilty Matrix*

In [106]:
probs_dict = {}

In [117]:
for words in complete_vocab:
    probs_dict[words] = (word_count[words]/n)*100

## *String Manipulation Functions*

***Delete Letter***

In [159]:
def delete(word):
    delete_l = [];
    m = len(word)
    for i in range(m):
        delete_l.append(word[:i] + word[i+1:])
    delete_l = set(delete_l)
    return delete_l

***Insert Letter***

In [160]:
def insert(word):
    insert_l = []
    letters = 'abcdefghijklmnopqrstuvwxyz'
    m = len(word) + 1
    for i in range(m):
        for j in range(26):
            temp = word[:i] + letters[j] + word[i:]
            insert_l.append(temp)
    insert_l = set(insert_l)
    return insert_l

***Replace Letter***

In [161]:
def replace(word):
    letters = 'abcdefghijklmnopqrstuvwxyz'
    replace_l = []
    m = len(word)
    for i in range(m):
        for j in range(26):
            if (word[i] != letters[j]):
                temp = word[:i] + letters[j] + word[i+1:]
                replace_l.append(temp)
    replace_l = set(replace_l)
    return replace_l
            

***Swap Letter***

In [162]:
def swap(word):
    swap_l = []
    m = len(word)
    for i in range(m-1):
        head   = word[:i]
        swap_1 = word[i]
        swap_2 = word[i+1]
        tail   = word[i+2:]
        temp   = head + swap_2 + swap_1 + tail
        swap_l.append(temp)
    swap_l = set(swap_l)
    return swap_l

## *Edit Function and Modal*

Edit Function combines all the above functions into one and returns all the possible manipulations

In [170]:
def edit(word,Delete = True,Swap = True, Insert = True, Replace = True):
    edits = []
    d = []
    i = []
    s = []
    r = []
    if(Delete):
        d = list(delete(word))
    if(Swap):
        s = list(swap(word))
    if(Insert):
        i = list(insert(word))
    if(Replace):
        r = list(replace(word))
    edits = d + s + i + r
    edits = set(edits)
    return edits
        

Suggest Function picks all the legit words from the list returned by the edit function

In [173]:
def Suggestions(word):
    temp_words = list(edit(word))
    m = len(temp_words)
    suggest = []
    for i in range(m):
        if (temp_words[i] in complete_vocab):
            suggest.append(temp_words[i])
    return suggest

### *Main Modaling Function*
<br>
Used for picking out the best possible Answer

In [211]:
def best_word(word):
    word = word.lower()
    if (word in complete_vocab):
        return word
    s = Suggestions(word)
    if(len(s) == 0):
        return 'No possible Word'
    m = len(s)
    best = 0
    auto_correct = ''
    for i in range(m):
        if(probs_dict[s[i]] > best):
            best = probs_dict[s[i]]
            auto_correct = s[i]
    return auto_correct
        

### Some Examples

In [220]:
a = 'What a beautifl scool'

In [221]:
b = a.split(' ')

In [222]:
c = []
for i in range(len(b)):
    c.append(best_word(b[i]))
print(a)
print(' '.join(word for word in c))

What a beautifl scool
what a beautiful school


In [242]:
d = 'bacj'

In [243]:
print(d)
print(best_word(d))

bacj
back


### Summary